### PR TITLE
Sync ``libasr`` from LPython/LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**
Corresponding LFortran PR - https://github.com/lfortran/lfortran/pull/705
Corresponding LPython PR - https://github.com/lcompilers/lpython/pull/1091

### CODE REVIEW

**Comment by czgdp1807** [comment]
@certik This and other sync PRs are ready.

File: src/libasr/CMakeFiles/CMakeDirectoryInformation.cmake [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# CMAKE generated file: DO NOT EDIT!
+# Generated by "Unix Makefiles" Generator, CMake Version 3.23
```
Comment (awvwgk) [inline src/libasr/CMakeFiles/CMakeDirectoryInformation.cmake]: Should this be checked in to the repository? Looks like the whole CMakeFiles directory is checked in on accident.

Comment (czgdp1807) [reply src/libasr/CMakeFiles/CMakeDirectoryInformation.cmake]: Ah! I did `git clean -fdx` before committing. Still this made into the commit. Removing.

Comment (czgdp1807) [reply src/libasr/CMakeFiles/CMakeDirectoryInformation.cmake]: Fixed.

**Comment by rebcabin** [comment]
Waiting for Ondrej for merge. I'm unblocked for my test-gen experiments. TY Gagandeep!

**Comment by rebcabin** [comment]
I'm kinda lost here amongst various PRs. The branch I'm looking at has no entry for "SymbolTable" in ASR.asdl. Please have a look at my comment in another PR: https://github.com/lcompilers/lpython/pull/1091#issuecomment-1238528009 .
I'm blocked on writing a clojure.spec for https://github.com/lcompilers/lpython/blob/main/tests/reference/asr-expr_01-211000e.stdout ., attempt here https://nextjournal.com/a/FnEMAE1Pkx2vWFMH6NVpJX/edit. 

I can infer from the example expr_01-211000e that a SymbolTable is a list with three elements: the tag symbol `SymbolTable`, an integer of unknown semantics, and a hash-map of key-value pairs, with each value recursing back into the adsl, as with a `Variable` or a `Function`. This must be written down in the adsl somewhere, but which version of the ADSL? I can't figure out how to set up a pull of the lpython repo where the examples in the tests/reference subdir conform to the speck in the https://github.com/lcompilers/lpython/blob/84a073ce44a9a74213a4ac5648ee783bd38fc90f/src/libasr/ASR.asdl

**Comment by czgdp1807** [comment]
@certik Could you please merge this as well? I don’t have access to do that.

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/lcompilers/libasr/pull/7

**Pull request closed without merging**